offchain/eth-input-reader/src/setup.rs

```diff
@@ -1,88 +1,108 @@
+pub type InputProvider = Provider<RetryClient<Http>>;
 
-pub async fn create_state_server(
-    config: &SCConfig,
-) -> Result<
-    impl StateServer<InitialState = RollupsInitialState, State = RollupsState>
-        + BlockServer,
-    EthInputReaderError,
-> {
-    let channel = Channel::from_shared(config.grpc_endpoint.to_owned())
-        .context(ChannelSnafu)?
-        .connect()
-        .await
-        .context(ConnectSnafu)?;
+pub async fn create_provider(
+    config: &EthInputReaderConfig,
+) -> Result<Arc<InputProvider>, EthInputReaderError> {
+    let http = Http::new(
+        Url::parse(&config.bh_config.http_endpoint).context(ParseSnafu)?,
+    );
+
+    let retry_client =
+        RetryClient::new(http, Box::new(HttpRateLimitRetryPolicy), 10, 1000);
+
+    let provider = Provider::new(retry_client);
 
-    Ok(GrpcStateFoldClient::new_from_channel(channel))
+    Ok(Arc::new(provider))
 }
 
-pub async fn create_block_subscription(
-    client: &impl BlockServer,
-    confirmations: usize,
+pub async fn create_block_subscriber(
+    config: &EthInputReaderConfig,
+    provider: Arc<InputProvider>,
+) -> Result<Arc<BlockSubscriber<InputProvider>>, EthInputReaderError> {
+    Ok(Arc::new(
+        BlockSubscriber::start(
+            Arc::clone(&provider),
+            config.bh_config.ws_endpoint.to_owned(),
+            config.bh_config.block_timeout,
+            config.bh_config.max_depth,
+        )
+        .await
+        .expect("should create block history"),
+    ))
+}
+
+pub async fn create_subscription(
+    block_subscriber: Arc<BlockSubscriber<InputProvider>>,
+    subscription_depth: usize,
 ) -> Result<
-    impl Stream<Item = Result<BlockStreamItem, StateServerError>>
-        + Send
+    impl Stream<Item = Result<BlockStreamItem, SubscriptionError<InputProvider>>>
         + std::marker::Unpin,
     EthInputReaderError,
 > {
-    let s = client
-        .subscribe_blocks(confirmations)
+    Ok(block_subscriber
+        .subscribe_new_blocks_at_depth(subscription_depth)
         .await
-        .context(StateServerSnafu)?;
-
-    let s = {
-        use futures::StreamExt;
-        s.ready_chunks(BUFFER_LEN)
-    };
+        .expect("should subscribe new blocks"))
+}
 
-    let s = s.filter_map(
-        |mut x| {
-            if x.len() == BUFFER_LEN {
-                None
-            } else {
-                x.pop()
-            }
-        },
+pub async fn create_env(
+    config: &EthInputReaderConfig,
+    provider: Arc<InputProvider>,
+    block_archive: Arc<BlockArchive<InputProvider>>,
+) -> Result<
+    Arc<StateFoldEnvironment<InputProvider, Mutex<UserData>>>,
+    EthInputReaderError,
+> {
+    let env = StateFoldEnvironment::new(
+        provider,
+        Some(block_archive),
+        config.sf_config.safety_margin,
+        config.sf_config.genesis_block,
+        config.sf_config.query_limit_error_codes.clone(),
+        config.sf_config.concurrent_events_fetch,
+        10000,
```

Comment:
What is this magic number? We should at least create a local `const magic_number = 10000`.

Reply:
I got this magic number directly from state-server, where is just as magical, you can check on the state-server/src/lib.rs at line 92. I checked the StateFoldEnvironment and copied the comment from there above the const inside the setup.